File: client/Assets/Ronriku/Scripts/Domain/Figures/Figure.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace Ronriku.Domain.Figures
{
    public enum FigureRarity
    {
        Common = 0,
        Rare = 1,
        Epic = 2,
        Legendary = 3
    }

    /// <summary>
    /// A voxel person: Size (x) × Size (y, front = 0) × 2·Size (z, up). Voxel values 0 = empty,
    /// 1–15 = index into <see cref="Palette"/> (RGB, 15 entries).
    /// </summary>
    public sealed class Figure
    {
        public const int PaletteSize = 15;

        public int Size { get; }
        public int Height => Size * 2;
        public uint[] Palette { get; }
        public byte[] Voxels { get; }
        public string Name { get; }
        public FigureRarity Rarity { get; }
        public FigureTraits Traits { get; }

        public Figure(int size, uint[] palette, byte[] voxels, string name, FigureRarity rarity, FigureTraits traits)
        {
            if (size < 3 || size > 5) throw new ArgumentOutOfRangeException(nameof(size));
            if (palette == null || palette.Length != PaletteSize) throw new ArgumentException("Palette must have 15 colours.");
            if (voxels == null || voxels.Length != size * size * size * 2) throw new ArgumentException("Voxel count mismatch.");
            Size = size;
            Palette = palette;
            Voxels = voxels;
            Name = name ?? string.Empty;
            Rarity = rarity;
            Traits = traits;
        }

        public int Index(int x, int y, int z) => x + Size * (y + Size * z);
        public byte Get(int x, int y, int z) => Voxels[Index(x, y, z)];

        public int FilledCount
        {
            get
            {
                int count = 0;
                foreach (byte v in Voxels) if (v != 0) count++;
                return count;
            }
        }

        /// <summary><c>RF1.{S}.{palette 90 hex}.{base64url nibble-packed voxels}</c>. See docs/PLAN_V2.md §5.</summary>
        public string Encode()
        {
            var sb = new StringBuilder("RF1.");
            sb.Append(Size).Append('.');
            foreach (uint rgb in Palette) sb.Append((rgb & 0xFFFFFF).ToString("X6"));
            sb.Append('.');
            var packed = new byte[(Voxels.Length + 1) / 2];
            for (int i = 0; i < Voxels.Length; i++)
            {
                int nibble = Voxels[i] & 0xF;
                if (i % 2 == 0) packed[i / 2] |= (byte)(nibble << 4);
                else packed[i / 2] |= (byte)nibble;
            }
            sb.Append(Convert.ToBase64String(packed).TrimEnd('=').Replace('+', '-').Replace('/', '_'));
            return sb.ToString();
        }

        /// <summary>Decodes geometry and palette. Name, rarity and traits are not part of the encoding.</summary>
        public static Figure Decode(string encoding)
        {
            string[] parts = (encoding ?? string.Empty).Split('.');
            if (parts.Length != 4 || parts[0] != "RF1") throw new FormatException("Not an RF1 figure.");
            int size = int.Parse(parts[1]);
            if (size < 3 || size > 5 || parts[2].Length != PaletteSize * 6) throw new FormatException("Bad figure header.");
            var palette = new uint[PaletteSize];
            for (int i = 0; i < PaletteSize; i++) palette[i] = Convert.ToUInt32(parts[2].Substring(i * 6, 6), 16);
            string b64 = parts[3].Replace('-', '+').Replace('_', '/');
            b64 = b64.PadRight(b64.Length + (4 - b64.Length % 4) % 4, '=');
            byte[] packed = Convert.FromBase64String(b64);
            var voxels = new byte[size * size * size * 2];
            if (packed.Length != (voxels.Length + 1) / 2) throw new FormatException("Voxel payload length mismatch.");
            for (int i = 0; i < voxels.Length; i++)
                voxels[i] = (byte)(i % 2 == 0 ? packed[i / 2] >> 4 : packed[i / 2] & 0xF);
            return new Figure(size, palette, voxels, string.Empty, FigureRarity.Common, default);
        }

        /// <summary>MagicaVoxel .vox (version 150): SIZE, XYZI and RGBA chunks. Palette index n maps to vox colour n.</summary>
        public byte[] ToVox()
        {
            var voxels = new List<byte[]>();
            for (int z = 0; z < Height; z++)
            for (int y = 0; y < Size; y++)
            for (int x = 0; x < Size; x++)
            {
                byte v = Get(x, y, z);
                // MagicaVoxel's y axis points away from the viewer; our y = 0 is the front.
                if (v != 0) voxels.Add(new[] { (byte)x, (byte)(Size - 1 - y), (byte)z, v });
            }

            using var stream = new MemoryStream();
            using var w = new BinaryWriter(stream);
            byte[] Chunk(string id, byte[] content)
            {
                using var m = new MemoryStream();
                using var cw = new BinaryWriter(m);
                cw.Write(Encoding.ASCII.GetBytes(id));
                cw.Write(content.Length);
                cw.Write(0);
                cw.Write(content);
                return m.ToArray();
            }

            byte[] size;
            using (var m = new MemoryStream())
            using (var cw = new BinaryWriter(m)) { cw.Write(Size); cw.Write(Size); cw.Write(Height); size = m.ToArray(); }
            byte[] xyzi;
            using (var m = new MemoryStream())
            using (var cw = new BinaryWriter(m))
            {
                cw.Write(voxels.Count);
                foreach (byte[] v in voxels) cw.Write(v);
                xyzi = m.ToArray();
            }
            var rgba = new byte[256 * 4];
            for (int i = 0; i < 255; i++)
            {
                uint rgb = i < PaletteSize ? Palette[i] : 0x808080u;
                rgba[i * 4] = (byte)(rgb >> 16);
                rgba[i * 4 + 1] = (byte)(rgb >> 8);
                rgba[i * 4 + 2] = (byte)rgb;
                rgba[i * 4 + 3] = 255;
            }

            byte[] children;
            using (var m = new MemoryStream())
            {
                foreach (byte[] c in new[] { Chunk("SIZE", size), Chunk("XYZI", xyzi), Chunk("RGBA", rgba) }) m.Write(c, 0, c.Length);
                children = m.ToArray();
            }
            w.Write(Encoding.ASCII.GetBytes("VOX "));
            w.Write(150);
            w.Write(Encoding.ASCII.GetBytes("MAIN"));
            w.Write(0);
            w.Write(children.Length);
            w.Write(children);
            return stream.ToArray();
        }
    }

    public readonly struct FigureTraits
    {
        public readonly int Skin, Outfit, Accent, Hair, Pants, Shoes, Headwear, Eyes, OutfitStyle, Accessory;
        public readonly bool Glow;

        public FigureTraits(int skin, int outfit, int accent, int hair, int pants, int shoes, int headwear, int eyes,
            int outfitStyle, int accessory, bool glow)
        {
            Skin = skin; Outfit = outfit; Accent = accent; Hair = hair; Pants = pants; Shoes = shoes;
            Headwear = headwear; Eyes = eyes; OutfitStyle = outfitStyle; Accessory = accessory; Glow = glow;
        }
    }
}

File: packages/core/src/figures/figure.ts
/**
 * Port of Ronriku.Domain.Figures.Figure (C#). A voxel person: size (x) × size (y, front = 0) × 2·size
 * (z, up). Voxel 0 = empty, 1–15 = index into `palette` (15 RGB entries, 0xRRGGBB).
 */
export type FigureRarity = "Common" | "Rare" | "Epic" | "Legendary";
export const FIGURE_RARITIES: readonly FigureRarity[] = ["Common", "Rare", "Epic", "Legendary"];

export const PALETTE_SIZE = 15;

export interface FigureTraits {
  skin: number;
  outfit: number;
  accent: number;
  hair: number;
  pants: number;
  shoes: number;
  headwear: number;
  eyes: number;
  outfitStyle: number;
  accessory: number;
  glow: boolean;
}

export class Figure {
  readonly size: number;
  readonly palette: number[];
  readonly voxels: Uint8Array;
  readonly name: string;
  readonly rarity: FigureRarity;
  /** Undefined for decoded figures (traits are not part of the encoding). */
  readonly traits?: FigureTraits;

  constructor(size: number, palette: number[], voxels: Uint8Array, name = "", rarity: FigureRarity = "Common", traits?: FigureTraits) {
    if (!Number.isInteger(size) || size < 3 || size > 5) throw new RangeError("size must be 3..5");
    if (palette.length !== PALETTE_SIZE) throw new Error("Palette must have 15 colours.");
    if (voxels.length !== size * size * size * 2) throw new Error("Voxel count mismatch.");
    this.size = size;
    this.palette = palette;
    this.voxels = voxels;
    this.name = name;
    this.rarity = rarity;
    this.traits = traits;
  }

  get height(): number {
    return this.size * 2;
  }

  index(x: number, y: number, z: number): number {
    return x + this.size * (y + this.size * z);
  }

  get(x: number, y: number, z: number): number {
    return this.voxels[this.index(x, y, z)];
  }

  get filledCount(): number {
    let count = 0;
    for (const v of this.voxels) if (v !== 0) count++;
    return count;
  }

  /** `RF1.{S}.{palette 90 hex}.{base64url nibble-packed voxels}`. See docs/PLAN_V2.md §5. */
  encode(): string {
    const hex = this.palette.map((rgb) => (rgb & 0xffffff).toString(16).toUpperCase().padStart(6, "0")).join("");
    const packed = new Uint8Array((this.voxels.length + 1) >> 1);
    for (let i = 0; i < this.voxels.length; i++) {
      const nibble = this.voxels[i] & 0xf;
      packed[i >> 1] |= i % 2 === 0 ? nibble << 4 : nibble;
    }
    return `RF1.${this.size}.${hex}.${toBase64Url(packed)}`;
  }

  /** Decodes geometry and palette. Name, rarity and traits are not part of the encoding. */
  static decode(encoding: string): Figure {
    const parts = (encoding ?? "").split(".");
    if (parts.length !== 4 || parts[0] !== "RF1") throw new Error("Not an RF1 figure.");
    if (!/^[345]$/.test(parts[1]) || !/^[0-9A-Fa-f]{90}$/.test(parts[2])) throw new Error("Bad figure header.");
    if (!/^[A-Za-z0-9_-]*$/.test(parts[3])) throw new Error("Bad voxel payload.");
    const size = Number(parts[1]);
    const palette: number[] = [];
    for (let i = 0; i < PALETTE_SIZE; i++) palette.push(parseInt(parts[2].substring(i * 6, i * 6 + 6), 16));
    const packed = fromBase64Url(parts[3]);
    const voxels = new Uint8Array(size * size * size * 2);
    if (packed.length !== (voxels.length + 1) >> 1) throw new Error("Voxel payload length mismatch.");
    for (let i = 0; i < voxels.length; i++) voxels[i] = i % 2 === 0 ? packed[i >> 1] >> 4 : packed[i >> 1] & 0xf;
    return new Figure(size, palette, voxels);
  }

  /** MagicaVoxel .vox (version 150): SIZE, XYZI and RGBA chunks. Palette index n maps to vox colour n. */
  toVox(): Uint8Array {
    const entries: number[][] = [];
    for (let z = 0; z < this.height; z++)
      for (let y = 0; y < this.size; y++)
        for (let x = 0; x < this.size; x++) {
          const v = this.get(x, y, z);
          // MagicaVoxel's y axis points away from the viewer; our y = 0 is the front.
          if (v !== 0) entries.push([x, this.size - 1 - y, z, v]);
        }

    const sizeContent = int32s(this.size, this.size, this.height);
    const xyzi = new Uint8Array(4 + entries.length * 4);
    new DataView(xyzi.buffer).setInt32(0, entries.length, true);
    entries.forEach((e, i) => xyzi.set(e, 4 + i * 4));
    const rgba = new Uint8Array(256 * 4);
    for (let i = 0; i < 255; i++) {
      const rgb = i < PALETTE_SIZE ? this.palette[i] : 0x808080;
      rgba[i * 4] = (rgb >> 16) & 0xff;
      rgba[i * 4 + 1] = (rgb >> 8) & 0xff;
      rgba[i * 4 + 2] = rgb & 0xff;
      rgba[i * 4 + 3] = 255;
    }
    const children = concat(chunk("SIZE", sizeContent), chunk("XYZI", xyzi), chunk("RGBA", rgba));
    return concat(ascii("VOX "), int32s(150), ascii("MAIN"), int32s(0, children.length), children);
  }
}

const B64 = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

/** base64url without padding (no Buffer dependency, so it also runs in browsers and Deno). */
export function toBase64Url(bytes: Uint8Array): string {
  let out = "";
  for (let i = 0; i < bytes.length; i += 3) {
    const n = (bytes[i] << 16) | ((bytes[i + 1] ?? 0) << 8) | (bytes[i + 2] ?? 0);
    const chars = Math.min(4, Math.ceil(((bytes.length - i) * 8) / 6));
    for (let c = 0; c < chars; c++) out += B64[(n >> (18 - c * 6)) & 63];
  }
  return out;
}

export function fromBase64Url(text: string): Uint8Array {
  const out: number[] = [];
  let buffer = 0,
    bits = 0;
  for (const ch of text) {
    const v = B64.indexOf(ch);
    if (v < 0) throw new Error("Bad base64url character.");
    buffer = ((buffer << 6) | v) & 0xffffff;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      out.push((buffer >> bits) & 0xff);
    }
  }
  return Uint8Array.from(out);
}

function ascii(text: string): Uint8Array {
  return Uint8Array.from(text, (c) => c.charCodeAt(0));
}

function int32s(...values: number[]): Uint8Array {
  const out = new Uint8Array(values.length * 4);
  const view = new DataView(out.buffer);
  values.forEach((v, i) => view.setInt32(i * 4, v, true));
  return out;
}

function chunk(id: string, content: Uint8Array): Uint8Array {
  return concat(ascii(id), int32s(content.length, 0), content);
}

function concat(...parts: Uint8Array[]): Uint8Array {
  const out = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let offset = 0;
  for (const p of parts) {
    out.set(p, offset);
    offset += p.length;
  }
  return out;
}
